File: backend/app/src/models/blueprints/Factory.py
from .Building import Building
from ..population.Worker import Worker; from ..population.Builder import Builder; from ..population.WarehouseWorker import WarehouseWorker; from ..population.Baker import Baker;
from ..population.Woodcutter import Woodcutter; from ..population.Miner import Miner; from ..population.Blacksmith import Blacksmith


class Factory(Building):
    build_options: dict =   {
                                "Builders Hut": Builder,
                                "Warehouse": WarehouseWorker,
                                "Bakery": Baker,
                                "Sawmill": Woodcutter,
                                "Gold Mine": Miner,
                                "Iron Mine": Miner,
                                "Forge": Blacksmith,
                            }
    
    def __init__(self, name: str, worker_slots: int = 0, workers = [], active: bool = True):
        super().__init__(name=name, active=active)
        self.worker_slots: int = worker_slots
        self.__workers: list = workers
    
    def fetch_building_state(self) -> dict:
        building_state: dict = super().fetch_building_state()
        building_state.update({
            "worker_slots": self.worker_slots,
            "workers": self.__workers
        })
        
        return building_state
    
    def get_workers(self, dump) -> list:
        return  { "success": True, "workers": self.__workers }
    
    def __set_workers(self, worker: Builder, increase: bool = True) -> None:
        if increase:
            self.__workers.append(worker)
        else:
            self.__workers.remove(worker)
    
    def add_worker(self, worker: Builder) -> None:
        self.__set_workers(worker)
        return  { "success": True }
    
    def remove_worker(self, worker: Builder) -> None:
        self.__set_workers(worker, increase=False)
        return  { "success": True }
    
    
    def convert_worker_to_craftsman(self, building: str, worker_id, settlement) -> bool:
        worker = settlement.select_population(worker_id)["target"]
        
        if isinstance(worker, Worker):
            new_craftsman = Factory.build_options[building](worker.name)
            settlement.remove_population(worker)
            settlement.set_free_workers(worker, False)
            del worker
            settlement.add_population(new_craftsman)
            
            if building == "Builders hut":
                settlement.add_free_builder(new_craftsman)
            
            return { 
                    "success": True, "message": f"{new_craftsman.name} wurde zum Baumeister!",
                    "update": {
                            "action": "Convert Worker",
                            "settlement_id": settlement.get_settlement_id(),
                            "old_population_id": worker_id,
                            "population_id": new_craftsman.get_population_id(),
                            "name": new_craftsman.name,
                            "profession": new_craftsman.profession,
                            "alive": new_craftsman.alive,
                            "employed": new_craftsman.employed,
                            "field_of_work": new_craftsman.field_of_work,
                            "working": new_craftsman.working,
                            "production": new_craftsman.production
                        } 
                    }

        return { "success": False, "message": f"{worker.name} konnte nicht zum Baumeister werden!" }

    # Method that takes workers etc as parameters to occupy them during work !?

File: backend_refactoring/CSharpApi/Models/BluePrints/Locations/Factory.cs
using CSharpApi.Models.BluePrints.Beings;

namespace CSharpApi.Models.BluePrints.Locations
{
    public class Factory : BuildingState
    {
        public Worker? BuildOptions(string buildingType, string name)
        { return buildingType switch {
             "Builders Hut" => new Builder(name),
             "Warehouse" => new WarehouseWorker(name),
             "Bakery" => new Baker(name),
             "Sawmill"=> new WoodCutter(name),
             "Gold Mine" => new Miner(name),
             "Iron Mine" => new Miner(name),
             "Forge" => new BlackSmith(name),
             _ => null
            };
        }
        
        private int WorkerSlots { get; set; }

        private List<Worker> Workers { get; set; } = [];

        public Factory(string buildingName, bool active = true, int workerSlots = 0, List<Worker> workers = null) : base(buildingName, active)
        {
            WorkerSlots = workerSlots;
            Workers = workers;
        }

        public Dictionary<string, object> FetchBuildingState()
        {
            Dictionary<string, object> buildingState = base.FetchBuildingState();
            buildingState.TryAdd("worker_slots", WorkerSlots);
            buildingState.TryAdd("workers", Workers);

            return buildingState;
        }

        public Dictionary<string, object> ConvertWorkerToCraftsman(string buildingName, int workerId, Settlement settlement)
        {
            Worker? worker = (Worker) settlement.SelectInstancePopulation(workerId)["target"];

            if (worker is not null)
            {
                var newCraftsman = BuildOptions(buildingName, worker.Name);
                settlement.Population.Remove(worker);
                settlement.SetInstancePopulation(worker, false);
                worker = null;
                if (newCraftsman is null) return new() { { "success", false}, { "reason", "Building Option Not Found" } };
                settlement.Population.Add(newCraftsman);

                if (buildingName.Equals("Builders Hut"))
                {
                    settlement.FreeBuilders.Add((Builder)newCraftsman);
                }

                return new Dictionary<string, object> {
                    { "success", true },
                    { "message", $"{newCraftsman.Name} wurde zum Baumeister!" },
                    {"update", new Dictionary<string, object> {
                        {"action", "Convert Worker" },
                        { "settlement_id",  settlement.SettlementId },
                        { "old_population_id", workerId },
                        { "population_id", newCraftsman.Id },
                        { "name", newCraftsman.Name },
                        {"profession", newCraftsman.Profession },
                        { "alive", newCraftsman.Alive },
                        { "employed", newCraftsman.Employed },
                        { "field_of_work", newCraftsman.FieldOfWork },
                        { "working", newCraftsman.Working },
                        { "production", newCraftsman.Production } }
                    }
                };
            }

            return new Dictionary<string, object>
            {
                { "success", false },
                { "message", $"Worker mit ID: {workerId} konnte nicht zum Baumeister werden!" }
            };
        }

        public Dictionary<string, object> GetWorker()
        {
            return new Dictionary<string, object> { { "success", true }, { "workers", Workers } };
        }


        /// <summary>
        /// Does what the name suggests "AddOrRemoveWorkers".
        /// Defaults to Adding a Worker.
        /// To Remove a Worker set <paramref name="increase"/> false.
        /// </summary>
        /// <param name="worker">The type of Worker to Add or Remove.</param>
        /// <param name="increase">The operation to perform.</param>
        /// <returns>Returns Dictionary @string, @bool.</returns>
        public Dictionary<string, bool> AddOrRemoveWorkers(Builder worker, bool increase = true)
        {
            switch (increase)
            {
                case false:
                    Workers.Remove(worker);
                    break;
                case true:
                    Workers.Add(worker);
                    break;
            }

            return new Dictionary<string, bool> { { "success", true } };

        }
    }
}
